engine/common/synchronization/engine_spam_test.go

```diff
@@ -0,0 +1,259 @@
+				}
+
+				// if request height is higher than local finalized, we should not respond
+				req.Height = ss.head.Height + 1
+
+				ss.core.On("HandleHeight", ss.head, req.Height)
+				ss.core.On("WithinTolerance", ss.head, req.Height).Return(false)
+				ss.con.AssertNotCalled(ss.T(), "Unicast", mock.Anything, mock.Anything)
+
+				// maybe function calls that might or might not occur over the course of the load test
+				ss.core.On("ScanPending", ss.head).Return([]chainsync.Range{}, []chainsync.Batch{}).Maybe()
+				ss.con.On("Multicast", mock.Anything, mock.Anything, mock.Anything, mock.Anything).Return(nil).Maybe()
+
+				// count misbehavior reports over the course of a load test
+				ss.con.On("ReportMisbehavior", mock.Anything).Return(mock.Anything).Maybe().Run(
+					func(args mock.Arguments) {
+						misbehaviorsCounter++
+					},
+				)
+				ss.e.spamDetectionConfig.syncRequestProbability = loadGroup.syncRequestProbabilityFactor
+				require.NoError(ss.T(), ss.e.Process(channels.SyncCommittee, originID, req))
+			}
+
+			// check function call expectations at the end of the load test; otherwise, load test would take much longer
+			ss.core.AssertExpectations(ss.T())
+			ss.con.AssertExpectations(ss.T())
+
+			// check that correct range of misbehavior reports were generated
+			// since we're using a probabilistic approach to generate misbehavior reports, we can't guarantee the exact number,
+			// so we check that it's within an expected range
+			ss.T().Logf("misbehaviors counter after load test: %d (expected lower bound: %d expected upper bound: %d)", misbehaviorsCounter, loadGroup.expectedMisbehaviorsLower, loadGroup.expectedMisbehaviorsUpper)
+			assert.GreaterOrEqual(ss.T(), misbehaviorsCounter, loadGroup.expectedMisbehaviorsLower)
+			assert.LessOrEqual(ss.T(), misbehaviorsCounter, loadGroup.expectedMisbehaviorsUpper)
+
+			misbehaviorsCounter = 0 // reset counter for next subtest
+		})
+	}
+}
+
+// TestLoad_Process_RangeRequest_SometimesReportSpam is a load test that ensures that a misbehavior report is generated
+// an appropriate range of times when the base probability factor and range are set to different values.
+func (ss *SyncSuite) TestLoad_Process_RangeRequest_SometimesReportSpam() {
+	ctx, cancel := irrecoverable.NewMockSignalerContextWithCancel(ss.T(), context.Background())
+	ss.e.Start(ctx)
+	unittest.AssertClosesBefore(ss.T(), ss.e.Ready(), time.Second)
+	defer cancel()
+
+	load := 1000
+
+	type loadGroup struct {
+		rangeRequestBaseProb      float32
+		expectedMisbehaviorsLower int
+		expectedMisbehaviorsUpper int
+		fromHeight                uint64
+		toHeight                  uint64
+	}
+
+	loadGroups := []loadGroup{}
+
+	// using a very small range (1) with a 10% base probability factor, expect to almost never get misbehavior report, about 0.003% of the time (3 in 1000 requests)
+	// expected probability factor: 0.1 * ((10-9) + 1)/64 = 0.003125
+	loadGroups = append(loadGroups, loadGroup{0.1, 0, 15, 9, 10})
```

Comment:
It seems that the `expectedMisbehaviorsUpper` and `expectedMisbehaviorsLower` are set approximately to shape a valid range where the expected value falls, i.e., we expected 3 in 1000. Hence, we shape the range as (0, 15). If this is a valid interpretation, please explain it in the comments.

Reply:
added godoc to explain loadGroup fields and how load tests work

3920b4d3ca86ad71fd531fb1b704ad7023f60a02